src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala

```diff
@@ -115,6 +115,35 @@ trait ClusterComponent extends LeoComponent {
       }
     }
 
+    def listByLabels(labelMap: Map[String, String]): DBIO[Seq[Cluster]] = {
+      val query = if (labelMap.isEmpty) {
+        clusterQueryWithLabels
+      } else {
+        // The trick is to find all clusters that have _at least_ all the labels in labelMap.
+        // In other words, for a given cluster, the labels provided in the query string must be
+        // a subset of its labels in the DB. The following SQL achieves this:
+        //
+        // select c.*, l.*
+        // from cluster c
+        // left join label l on l.clusterId = c.id
+        // where (
+        //   select count(*) from label
```

Comment:
My "inner select is slow" alarm bell just went off. But I don't think from the code below that that's what Slick will generate here -- just a complicated WHERE clause. Am I right?

Reply:
It's an inner select:
```
select x2.`id`, x2.`clusterName`, x2.`googleId`, x2.`googleProject`, x2.`googleServiceAccount`, x2.`googleBucket`, x2.`operationName`, x2.`status`, x2.`hostIp`, x2.`createdDate`, x2.`destroyedDate`, x2.`jupyterExtensionUri`, x3.`clusterId`, x3.`clusterId`, x3.`key`, x3.`value` from `CLUSTER` x2 left outer join `LABEL` x3 on x2.`id` = x3.`clusterId` where (select count(1) from `LABEL` where (`clusterId` = x2.`id`) and (((`key` = 'foo') and (`value` = 'bar')) or ((`key` = 'vcf') and (`value` = 'no')))) = 2
```

Reply:
OK. Don't have to do anything now but let's keep an eye on this. We've had nasty performance issues with these before.